### DISCUSSION TITLE
LVGL clear screen

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#21170] [Q&A]
How to clear the screen?
even after reboot the screen keeps the content.
i tried:
```
lv.obj_clean(lv.scr_act());
BRY: Exception> 'attribute_error' - module 'lv' has no member 'obj_clean'
```

**Comment by s-hadinger** [comment]
At boot the splash screen is displayed and the screen is then cleared after 5 seconds if lvgl is not used. 

Can you describe your problem in more details?

**Reply by ChrisVonFrosta** [reply]
I deactivated the splash screen because it takes up too much time due to my many upload iterations.
I use the display with lvgl/berry.  I can draw objects. But i can't clear the screen content. If i  reboot i would like to clear the screen first- make it black/empty.

The LVGL documentation states that this is done using `lv.obj_clean(lv.scr_act());`.

**Reply by augea** [reply]
I have the same problem. 
I'm trying to write a program with LVGL.
I distributed some buttons on the main_scr.
Then I want to move the bottons a bit.
So I make the changes to the position and start the program again
The result is, the old position and new buttons are now on the screen.
I tried to clear the srceen with the command `lv.obj_clean(lv.scr_act())`
but the error message says:

```BRY: Exception> 'attribute_error' - module 'lv' has no member 'obj_clean'
stack traceback:
	<unknown source>: in function `init`
	<unknown source>: in function `main
```

Here is my code segent:	

```def init() # LVGM initialisieren falls noch nicht geschehen
        # Instanzenvariablen initialisieren
         # self.screen_initialized = false # Wurde LVGM schon initalisiert?
         self.stack = []   # Liste der definierten LVGM screens 
         self.hres  = nil  # Display native horizontale Auslösung
         self.vres = nil   # Display native vertikale Auslösung
         self.f20  = nil   # Font Variable 
         self.main_scr = nil
         # lokale Variablen der Init-Routine
        if !self.screen_initialized
            # Display-Hardware initialisieren
            lv.start()
            self.screen_initialized = true   # Screen wurde initialisiert
            self.hres = lv.get_hor_res()     # should be 480
            self.vres = lv.get_ver_res()     # should be 320
            self.f20 = lv.montserrat_font(20)  # load embedded Montserrat 20
            self.main_scr = lv.scr_act()     # default screen object
            self.set_bg(self.main_scr)       # Funktionsaufruf zum setzen des Hintergrund
            self.stack.push(self.main_scr)   # Aufnehmen des main_scr als erstes Element in den Screen stack
            print ('LVGM initlialized:',  self.screen_initialized) # Debug Print
            lv.obj_clean(lv.scr_act())
```

**Reply by sfromis** [reply]
As documented in the Berry [LVGL API Reference](https://tasmota.github.io/docs/LVGL_API_Reference/), the `lv_obj_clean` function is implemented as the object having a `clean()` method call.

Additionally, for development, you do have the mechanism of restarting the Berry VM (without rebooting all of Tasmota).
https://tasmota.github.io/docs/Berry/#iterate-without-rebooting

Already for other reasons, a build with `#define USE_BERRY_DEBUG` is recommended, but even without that, you have the command `BrRestart`.

**Reply by augea** [reply]
I've enabled the debug switch. However, the button is no longer displayed in the Berry console with Tasmota 14.5.0.3 (tasmota32). Restarting the console works with BrRestart, but it doesn't clear the screen. The LVGL buttons remain visible after restarting the VM. So I tried using Clear. Either I'm using it incorrectly (see code segemnt obove) or it doesn't work. The idea was to issue a Clear immediately after lv.start and fetching the active screen so that all widgets disappear. But neither method works. Only a complete Tasmota reboot clears the screen.

**Reply by sfromis** [reply]
I was only pointing out how the method call for clean was not "missing", as it is documented where it is. Not suggesting that it would be a solution for you.

Specifically, I'd not expect the `clean()` method (not `clear`) to communicate with the display about erasing (or clearing) contents, only to cleanup based internal objects already in LVGL. Hence my expectation would be that the `clean()` method would already not make sense "immediately after `lv.start`".

And some display types may have a buffer as part of the display controller, instead of depending on refresh from the host. In some situation I've even seen display content linger after restarting Tasmota, but I don't remember any details of that.

Do note that I'm not doing much at the LVGL level (mostly using HASPmota), and can't really tell you how to do things. And without demonstration code to recreate your issue, no basis for even making guesses.

I'm seeing the restart button for a long time, when having the debug option enabled. 
![image](https://github.com/user-attachments/assets/15256ad4-2c95-4390-b455-b8060252ebe6)

However, as LVGL is not part of Berry, I'm not sure that restarting the Berry VM should be "expected" to reinitialize displays at a "lower level".

**Reply by augea** [reply]
Well, after a BrRestart all widgets are still visible on the screen. 
If `lv.start()` would clear all LVGL resouces it would be fine for me. 
Can somebody confim this?

**Reply by s-hadinger** [reply]
No, `lv.start()` is just used to start LVGL the first time. Any other call to `lv.start()` does nothing.

LVGL has its own life, independent from Berry. So calling `BrRestart` does not clear anything in LVGL. Actually it is even worse, because some memory allocated by Berry may be used by LVGL. Freeing all this memory can cause LVGL to crash.

What are you trying to do?
- Refresh the screen?
- Clear everything you defined in LVGL
- Or just temporarily clear the screen to display its content back later? If this is what you want, there are ways, like creating a new screen object and swap the current screen. Or put a black rectangle in front of all objects (basically what is used for screen saving)

**Reply by augea** [reply]
Yes, I’m trying to clean resources allocated by LVGL.
During testing I have to restart the program several times but widgets are not cleared from screen without a complete ESP reboot.
I.e. if you set a button and change its position by changing and reentering the program you get the old button and the new button placed over each other.
See also `Berry restarts when deleting a not active screen (LVGL) #23327` (same project)
Im' looking for a way to release all and parts of the resources allocated by LVGL.

**Reply by s-hadinger** [reply]
The only way to release resources allocated by LVGL is to explicitly ask LVGL to delete all objects one by one. This means that you need to call `<obj>.delete()` on each object.

**Comment by augea** [comment]
I understand what you wrote. However, I'd like to ask again specifically. So, deleting the parent object (like the screen object) doesn't work, but rather requires deleting each object separately? That would be different from the LVGL documentation, and the solution is different in Berry on Tasmota.


`There is a common delete function for all object types. It deletes the object and all of its children.` 
see
`https://docs.lvgl.io/8.3/overview/object.html`

**Reply by s-hadinger** [reply]
Berry does a direct mapping of lvgl objects. As said this part is not deeply tested. So yes deleting the parent object should remove sub-objects in lvgl. But you need to be cautious of not leaving Berry references either or the memory won't be freed
